File: OPL/PhongVo_OPL_A1/A1.pl
# UMass Lowell
# Organization of Programming Languages (OPL)
# by Phong Vo - Spring 2019
# Assignment 1
# reference link: https://rosettacode.org/wiki/Permutations#Prolog

:- use_module(library(clpfd)). 	# using the clpfd library
 
permut_clpfd(L, N) :-
    length(L, N),
    L ins 1..N,
    all_different(L),
    label(L).

File: OPL/PhongVo_OPL_A1/A1.hs
-- UMass Lowell
-- Organization of Programming Languages (OPL)
-- by Phong Vo - Spring 2019
-- Assignment 1
-- reference link: https://rosettacode.org/wiki/Permutations#Haskell

module Main where
import Data.Char

permutations :: [a] -> [[a]]
permutations =
	foldr (\x ac -> ac >>= (fmap . ins x) <*> (enumFromTo 0 . length)) [[]]
	where
		ins x xs n =
		let (a, b) = splitAt n xs
		in a ++ x : b
 
main :: IO ()
main = print $ permutations[1, 2, 3]